src/runtime/rocm/rocm_module.cc

```diff
@@ -0,0 +1,278 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file rocm_module.cc
+ */
+#include "./rocm_module.h"
+
+#if TVM_ROCM_RUNTIME
+
+#include <tvm/runtime/registry.h>
+#include <hip/hip_runtime_api.h>
+#include <vector>
+#include <array>
+#include <string>
+#include <mutex>
+#include "./rocm_common.h"
+#include "../pack_args.h"
+#include "../thread_storage_scope.h"
+#include "../meta_data.h"
+#include "../file_util.h"
+
+namespace tvm {
+namespace runtime {
+
+// Module to support thread-safe multi-GPU execution.
+// hipModule_t is a per-GPU module
+// The runtime will contain a per-device module table
+// The modules will be lazily loaded
+class ROCMModuleNode : public runtime::ModuleNode {
+ public:
+  explicit ROCMModuleNode(std::string data,
+                          std::string fmt,
+                          std::unordered_map<std::string, FunctionInfo> fmap,
+                          std::string hip_source)
+      : data_(data), fmt_(fmt), fmap_(fmap), hip_source_(hip_source) {
+    std::fill(module_.begin(), module_.end(), nullptr);
+  }
+  // destructor
+  ~ROCMModuleNode() {
+    for (size_t i = 0; i < module_.size(); ++i) {
+      if (module_[i] != nullptr) {
+        ROCM_CALL(hipSetDevice(static_cast<int>(i)));
+        ROCM_DRIVER_CALL(hipModuleUnload(module_[i]));
+      }
+    }
+  }
+
+  const char* type_key() const final {
+    return "hip";
+  }
+
+  PackedFunc GetFunction(
+      const std::string& name,
+      const std::shared_ptr<ModuleNode>& sptr_to_self) final;
+
+  void SaveToFile(const std::string& file_name,
+                  const std::string& format) final {
+    std::string fmt = GetFileFormat(file_name, format);
+    std::string meta_file = GetMetaFilePath(file_name);
+    if (fmt == "cpp") {
+      CHECK_NE(hip_source_.length(), 0);
+      SaveMetaDataToFile(meta_file, fmap_);
+      SaveBinaryToFile(file_name, hip_source_);
+    } else {
+      CHECK_EQ(fmt, fmt_)
+          << "Can only save to format=" << fmt_;
+      SaveMetaDataToFile(meta_file, fmap_);
+      SaveBinaryToFile(file_name, data_);
+    }
+  }
+
+  void SaveToBinary(dmlc::Stream* stream) final {
+    stream->Write(fmt_);
+    stream->Write(fmap_);
+    stream->Write(data_);
+  }
+
+  std::string GetSource(const std::string& format) final {
```

Comment:
It is likely we won't have hip source as code are directly generated in llvm. We can have something similar to ptx though to get returned. Maybe it makes sense to remove hip_source_ field